iree/compiler/Conversion/LinalgToSPIRV/ConvertToGPUPass.cpp

```diff
@@ -428,35 +468,63 @@ static LogicalResult mapToProcessorsAndGuard(
   SmallVector<Value, 2> id(numLoops), count(numLoops);
   getGPUProcessorIdsAndCounts<GPUIdOp, GPUCountOp>(pLoopOp.getLoc(), rewriter,
                                                    numLoops, id, count);
-  return mapToProcessorsAndGuard(rewriter, pLoopOp, id);
+  return distributeSingleIterationPerProcessor(rewriter, pLoopOp, id,
+                                               generateGuard);
 }
 
 /// Distribute the scf.parallel to workgroups.
 static LogicalResult mapToWorkgroups(ConversionPatternRewriter &rewriter,
-                                     scf::ParallelOp pLoopOp) {
-  return mapToProcessor<gpu::BlockIdOp, gpu::GridDimOp>(rewriter, pLoopOp);
+                                     scf::ParallelOp pLoopOp,
+                                     bool useCyclicDistribution = false) {
+  if (useCyclicDistribution)
+    return distributeCycliclyToProcessors<gpu::BlockIdOp, gpu::GridDimOp>(
+        rewriter, pLoopOp);
+  return distributeSingleIterationPerProcessor<gpu::BlockIdOp, gpu::GridDimOp>(
+      rewriter, pLoopOp, false);
 }
 
 /// Distributes scf.parallel to workitems using local invocation ID.
 static LogicalResult mapToLocalInvocationId(ConversionPatternRewriter &rewriter,
                                             scf::ParallelOp pLoopOp) {
-  return mapToProcessorsAndGuard<gpu::ThreadIdOp, gpu::BlockDimOp>(rewriter,
-                                                                   pLoopOp);
+  return distributeSingleIterationPerProcessor<gpu::ThreadIdOp,
+                                               gpu::BlockDimOp>(rewriter,
+                                                                pLoopOp);
 }
 
 /// Distributes scf.parallel to workitems using global invocation ID. The GPU
 /// dialect doesn't have a direct operation to do this. This could be done using
 /// id = blockIdx * blockDim + gridIdx. count = blockDim * gridDim.
 static LogicalResult mapToGlobalInvocationId(
     ConversionPatternRewriter &rewriter, scf::ParallelOp pLoopOp) {
-  return mapToProcessorsAndGuard<GPUGlobalId, GPUGlobalCount>(rewriter,
-                                                              pLoopOp);
+  return distributeSingleIterationPerProcessor<GPUGlobalId, GPUGlobalCount>(
+      rewriter, pLoopOp);
 }
 
 //===----------------------------------------------------------------------===//
 // Pass and patterns.
 //===----------------------------------------------------------------------===//
 
+/// In some cases the iterations of the loops when partitioned to workgroups
+/// need to be distributed in a cyclic manner. The main use cases here is when
+/// the number of workgroups is constrained such that the number of iterations
+/// is greater than equal to number of processors (along any dimension). In
+/// those cases, distribute the iterations in a cyclic manner. This adds
+/// additional control flow, but isnt too detrimental to performance since they
```

Comment:
For my own education, is there a rule of thumb you have in mind for perf impact?

Reply:
Control flow within a kernel itself is really detrimental for performance if the branch is "divergent", i.e. different workitems within a subgroup take different paths at the branch node. If all the workitems take the same path, then the impact is lesser. It is non-zero since presence of branches typically means more register usage which can make you fall off a performance cliff if you end up spilling. So
1) No branching
2) Branching but all workitems within a subgroup take the same path
3) Branching with workitems within a subgroup taking different paths
is in increasing order of having adverse impact on performance.

Reply:
Yes. Scalarization is an important optimization to improve performance there. You typically see terms like "uniform", "dynamic uniform", and "non uniform" in Vulkan/SPIR-V land, which roughly maps the above 3 cases Mahesh said. Uniform (where we know at compile-time all lanes use the same value) is nice given that we can just use SALU to perform comparison and bypass a chunk of code directly without even engaging the VALUs. Dynamic uniform (where we cannot know at compile-time all lanes use the same value but at run time they are) is also helpful where in driver compilers we can also emit additional scalar comparisons to bypass code snippet if the lane mask register turns to be all 0s or all 1s. Non-uniform is the most difficult ones; you just need to engage VALUs and run both sides of the if statement.